include/storage/flash_map.h

```diff
@@ -237,6 +237,31 @@ int flash_area_has_driver(const struct flash_area *fa);
  */
 const struct device *flash_area_get_device(const struct flash_area *fa);
 
+/**
+ * Get the value expected to be read when accessing any erased
+ * flash byte.
+ *
+ * @param fa Flash area.
+ *
+ * @return Byte value of erase memory.
+ */
+uint8_t flash_area_erased_val(const struct flash_area *fa);
+
+/**
+ * Reads len bytes from off, and checks if the read data is erased.
+ *
+ * @param[in]  fa  Flash area
+ * @param[in]  off Offset relative from beginning of flash area to read
+ * @param[out] dst Buffer to store read data
+ * @param[in]  len Number of bytes to read
+ *
+ * @returns 1 if memory erased, 0 if non-erased,
```

Comment:
Since this API is mandated by MCUboot we have to do what it specifies.   (That should be mentioned in the description of the function.)

Ah.  I see.  In that case, MCUboot should be specifying what the function return value should be under all conditions.  Instead of using a function in common code that requires every back-end to define and declare it (where every backend appears to be `cypress` and `zephyr` and `sim`.

MCUboot failed to be clear, and there are other ports that Zephyr's implementation must behave consistently with. MCUboot  `docs/PORTING.md` currently has:

```
/*< Reads len bytes from off, and checks if the read data is erased. Returns
    1 if empty (that is containing erased value), 0 if not-empty, and -1 on
    failure. */
int     flash_area_read_is_empty(const struct flash_area *fa, uint32_t off,
                     void *dst, uint32_t len);
```
While the text there could support `len == 0` returning zero, all three current back-ends (cypress, zephyr, and sim) return 1 in that case.

So this function needs to do so too.

Reply:
Also the error return value is `-1`, nothing else.

Reply:
> While the text there could support `len == 0` returning zero, all three current back-ends (cypress, zephyr, and sim) return 1 in that case.

Seems that they took `Returns 1 if empty` literally, buffer with `len == 0` of data is indeed empty.

Reply:
> MCUboot failed to be clear, and there are other ports that Zephyr's implementation must behave consistently with. MCUboot `docs/PORTING.md` currently has:

Maybe we should propose changes to Mcuboot, make the documentation more clear and/or refactor it there with cooperation with other back-end providers.

Reply:
> Maybe we should propose changes to Mcuboot, make the documentation more clear and/or refactor it there with cooperation with other back-end providers.

You would also have to submit a PR to Mynewt, since the API was originally generated there. I don't expect much friction, but I would defer merging it to other committers.

Reply:
I don't think it's worth trying to change upstream.  I'd rather seek us go ahead with `flash_area_erased_value()` which is independently useful, and leave `flash_area_read_is_empty()` as something that exists only to conform to MCUboot's existing API.  Ideally that function would be conditional on a Kconfig signalling MCUboot support.

Reply:
Maybe we can change description of the return value in MCUBoot. MCUboottreads any negative value as an error.
Also name might be changed (not sure, @utzig will check that).
Will ensure consistence of behavior while `len==0`.

Reply:
> Maybe we can change description of the return value in MCUBoot. MCUboottreads any negative value as an error.

Makes sense, send PR!

> Also name might be changed (not sure, @utzig will check that).

This is used only by MCUboot and fcb/fcb2 in Mynewt so should be easy to change when we settle on some better naming convention. Btw, the `fcb` in Zephyr won't work on devices with flash erased to 0 in its current form.

Reply:
this function is not in scope anymore